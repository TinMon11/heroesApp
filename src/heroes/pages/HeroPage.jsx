import { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { GetHeroeById } from '../helpers'

export const HeroPage = () => {

    const {id} = useParams()
    const navigate = useNavigate()

    const hero = useMemo(()=> GetHeroeById(id), [id]);

    const onNavigateBack = ()=> {
        navigate(-1) //navega a la pagina anterior que estuvo
    }


    if (!hero) { 
        return <Navigate to ='/marvel'/>
    }

    return (
        <div className="row mt-5 mx-auto">
            <div className="col-4 animate__animated animate__fadeInLeft">
                <img 
                src={`/assets/heroes/${id}.jpg`}
                alt={hero.superhero}
                className='img-thumbnail' />
            </div>
            <div className="col-8 animate__animated animate__slideInRight">
                <h3>{hero.superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className='list-group-item'><b>Alter Ego: </b>{hero.alter_ego}</li>
                    <li className='list-group-item'><b>Publisher: </b>{hero.publisher}</li>
                    <li className='list-group-item'><b>First Appearence: </b>{hero.first_appearance}</li>
                </ul>

                <h5 className='mt-5'>Characters</h5>
                <p>{hero.characters}</p>

            <button className='btn btn-outline-primary'
            onClick={onNavigateBack}>
                Back
            </button>

            </div>

        </div>
    )
}
