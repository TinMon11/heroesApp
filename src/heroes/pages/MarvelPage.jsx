import React from 'react'
import { HeroList } from '../components'

export const MarvelPage = () => {
  return (
    <>
    <h1 className='text-center mt-2'>Marvel Page</h1>
    <hr />
    <HeroList publisher={'Marvel Comics'} />
    </>
  )
}
