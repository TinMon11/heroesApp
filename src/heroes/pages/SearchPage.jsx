import React from 'react'
import { useLocation, useNavigate } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import queryString from 'query-string'
import { HeroCard } from '../components'
import { GetHeroesByName } from '../helpers'

export const SearchPage = () => {

    const navigate = useNavigate()
    const location = useLocation()

    const { q = ''} = queryString.parse(location.search)

    const heroes = GetHeroesByName(q);

    const showSearch = (q.length === 0) // esto devuelve true o false
    const showError = (q.length > 0) && (heroes.length === 0) // esto devuelve true o false

    const { searchText, onInputChange } = useForm({
        searchText: q,
    })

    const onSearchSubmit = (event) => {
        event.preventDefault()
        if (searchText.trim().length <= 1) return;

        navigate(`?q=${searchText.toLowerCase().trim()}`)
    }


    return (
        <>
            <h1 className='text-center mt-2'>Search By Hero</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={onSearchSubmit}>
                        <input type="text"
                            placeholder='Hero Name'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={onInputChange} />
                    </form>
                    <button className='btn btn-outline-primary mt-2'>Search</button>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    
                    <div className="alert alert-primary animate__animated animate__fadeIn" 
                    style={{display: showSearch ? '' : 'none'}}>Search a Hero</div>

                    <div className="alert alert-danger animate__animated animate__fadeIn"
                    style={{display: showError ? '' : 'none'}}>Hero <b>{q}</b> not found</div>
                    
                    {heroes.map(
                        hero => (<HeroCard key={hero.id} {...hero} />))
                    }

                </div>
            </div>
        </>
    )
}
