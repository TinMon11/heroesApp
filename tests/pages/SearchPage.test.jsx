import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', ()=> ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}))

describe("Pruebas en Search Page", () => {

  beforeEach(()=> jest.clearAllMocks())

  test("Debe mostrarse correctamente con default values", () => {
    
    const {container} = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("Debe mostar Batman y el input con el query string", () => {
    
    
    const {container} = render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    );
    
    const input = screen.getByRole('textbox')
    expect(input.value).toBe('batman')

    const imagen = screen.getByRole('img')
    expect(imagen.src).toContain('batman')

    const alert = screen.getByLabelText('alert-danger')
    expect(alert.style.display).toEqual('none')

  });

  test('Debe mostrar error si no se encuentra el Hero (Batman123)', () => { 
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    );

    const alert = screen.getByLabelText('alert-danger')
    expect(alert.style.display).toBe('')

   })

   test('Debe llamar el Navigate a la nueva pantalla', () => { 

    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox')
    fireEvent.change(input, {target: {name: 'searchText', value: 'superman'}})
    
    const form = screen.getByRole('form')
    fireEvent.submit(form);

      expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman')


    })

});
