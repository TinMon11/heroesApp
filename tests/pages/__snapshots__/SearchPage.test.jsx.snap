// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Page Debe mostrarse correctamente con default values 1`] = `
<div>
  <h1
    class="text-center mt-2"
  >
    Search By Hero
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Hero Name"
          type="text"
          value=""
        />
      </form>
      <button
        class="btn btn-outline-primary mt-2"
      >
        Search
      </button>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        Hero 
        <b />
         not found
      </div>
    </div>
  </div>
</div>
`;
